action/binary.go

```diff
@@ -0,0 +1,185 @@
+	from := c.Args().Get(0)
+	to := c.Args().Get(1)
+
+	return s.binaryCopy(from, to, false)
+}
+
+// BinaryMove works like BinaryCopy but will remove (shred/wipe) the source
+// after a successfull copy. Mostly useful for securely moving secrets into
+// the store if they are no longer needed / wanted on disk afterwards
+func (s *Action) BinaryMove(c *cli.Context) error {
+	from := c.Args().Get(0)
+	to := c.Args().Get(1)
+
+	return s.binaryCopy(from, to, true)
+}
+
+// binaryCopy implements the control flow for copy and move. We support two
+// workflows:
+// 1. From the filesystem to the store
+// 2. From the store to the filesystem
+//
+// Copying secrets in the store must be done through the regular copy command
+func (s *Action) binaryCopy(from, to string, deleteSource bool) error {
+	switch {
+	case fsutil.IsFile(from) && fsutil.IsFile(to):
+		// copying from on file to another file is not supported
+		return fmt.Errorf("ambiquity detected. Only from or to can a file")
+	case s.Store.Exists(from) && s.Store.Exists(to):
+		// copying from one secret to another secret is not supported
+		return fmt.Errorf("ambiquity detected. Either from or to must be a file")
+	case fsutil.IsFile(from) && !fsutil.IsFile(to):
+		// if the source is a file the destination must no to avoid ambiquities
+		// if necessary this can be resolved by using a absolute path for the file
+		// and a relative one for the secret
+		if !strings.HasSuffix(to, BinarySuffix) {
+			to += BinarySuffix
+		}
+		// copy from FS to store
+		buf, err := ioutil.ReadFile(from)
+		if err != nil {
+			return err
+		}
+		if err := s.Store.Set(to, []byte(base64.StdEncoding.EncodeToString(buf)), fmt.Sprintf("Copied data from %s to %s", from, to)); err != nil {
+			return err
+		}
+		if deleteSource {
+			if err := s.binaryValidate(buf, to); err != nil {
+				return err
+			}
+			return fsutil.Shred(from, 8)
+		}
+		return nil
+	case !fsutil.IsFile(from):
+		// if the source is no file we assume it's a secret and to is a filename
+		// (which may already exist or not)
+		if !strings.HasSuffix(from, BinarySuffix) {
+			from += BinarySuffix
+		}
+		// copy from store to FS
+		buf, err := s.binaryGet(from)
+		if err != nil {
+			return err
+		}
+		if err := ioutil.WriteFile(to, buf, 0600); err != nil {
+			return err
+		}
+		if deleteSource {
+			if err := s.binaryValidate(buf, from); err != nil {
+				return err
+			}
+			return s.Store.Delete(from)
+		}
+		return nil
+	default:
+		return fmt.Errorf("ambiquity detected. Unhandled case. Please report a bug")
+	}
+}
+
+func (s *Action) binaryValidate(buf []byte, name string) error {
+	h := sha256.New()
+	_, _ = h.Write(buf)
+	fileSum := fmt.Sprintf("%x", h.Sum(nil))
+
+	h.Reset()
+
+	var err error
+	buf, err = s.binaryGet(name)
+	if err != nil {
+		return err
+	}
+	_, _ = h.Write(buf)
+	storeSum := fmt.Sprintf("%x", h.Sum(nil))
+
+	if fileSum != storeSum {
+		return fmt.Errorf("Hashsum mismatch (file: %s, store: %s)", fileSum, storeSum)
```

Comment:
Indeed, but we'll have to do a big round of error handling / propagation refactoring anyway so I'd like to postpone this. I don't want to introduce one kind of error propagation here that may differ from the way we decide it later.

But definitely a good point, we should keep this in mind.